Added dockerfile and fixed Axios client base URL

KianAttar opened the pull request:
### Pull Request Summary

This pull request introduces several important updates to enhance the configuration and deployment of the front-end application. The following changes have been made:

1. **Added Dockerfile**:
   - A `Dockerfile` has been added to configure the server environment for Vite to work effectively within a container. This ensures that the Vite development server operates correctly when the application is deployed in a containerized environment.

2. **Updated `.env` File Handling**:
   - The `.env` files have been removed from `.gitignore`, and a new `.env` file has been added for Vite configuration. 
   - Since frontend applications should not contain secrets, this `.env` file now includes configuration variables that are safe to be shared across different files. This setup allows us to manage configuration more effectively and keep secrets out of version control.
   - This change will help in later stages by allowing different sets of configuration variables for production environments without requiring changes to the codebase.

3. **Base URL Update**:
   - The base URL for Axios clients has been updated to utilize the environment variables specified in the `.env` file. This change ensures that the API endpoints are dynamically configured based on the environment, making the application more flexible and maintainable.

4. **Merged Updates from Remote Repository**:
   - Integrated the latest changes from the upstream repository to keep the fork in sync with the main project.


For more details on the Vite environment variable configuration, please refer to the Vite documentation: [Vite Env and Mode Guide](https://vitejs.dev/guide/env-and-mode).

### Additional Information for Development

For a more integrated development experience, you can work with the frontend by forking the repository available at [https://github.com/hassan-attar/CarX](https://github.com/hassan-attar/CarX). This repository includes both the backend and frontend as submodules and provides a complete stack setup using Docker Compose. 

Here's what you need to do:
- **Fork the Repository**: Fork [CarX](https://github.com/hassan-attar/CarX) to your own GitHub account.
- **Install Docker**: Ensure you have Docker installed on your local machine.
- **Build and start**: Run the provided commands in the `package.json` file to start the entire stack. To build and start, you can run `npm run build-start`. There are other commands to create fresh database or load sample data.

The containers stack are:
  - **Nginx Reverse Proxy**
  - **Frontend Application**
  - **Node API**
  - **Application Database**
  - **Authentication Database**

   All components run locally, and the React app supports hot reload. The source directory is mounted, so changes are reflected in real time, providing a great development experience.

Please reach out to me if you have further questions.

Commit:
added dockerfile

Signed-off-by: Hassan Attar <info@hassanattar.com>

Commit:
add server configuration for vite to work in container environment for dev

Signed-off-by: Hassan Attar <info@hassanattar.com>

Commit:
removed .env files from gitignore and added .env file for vite

frontend cannot have secrets. so some configuration variables that can be shared between files are
put into the .env file.
later on we can have different set of configuration for prod environment without changing the code itself.

the base URL for axios clients was updated to use that environment varibales.

for more info, please refer to vite docs at: https://vitejs.dev/guide/env-and-mode

Signed-off-by: Hassan Attar <info@hassanattar.com>

Commit:
Merge remote-tracking branch 'forked-repo/main'

Merging my suggested changes to the main repo.

Signed-off-by: Hassan Attar <info@hassanattar.com>

Commit:
removed .env files from dockeringore for the same reason; removed backend submodule, since we don't need it anymore

Signed-off-by: Hassan Attar <info@hassanattar.com>

amir-yari commented:
Hi Hassan,

Thank you for the detailed summary of the changes made in this pull request. The updates look promising and will definitely enhance our configuration and deployment processes. I have a few comments and questions:

1. **Dockerfile Addition:**
   - It's great to see the inclusion of a Dockerfile for the Vite development server. Could you provide a brief overview of the Dockerfile's structure or any specific configurations you included to ensure Vite operates effectively within the container?

2. **Updated .env File Handling:**
   - I understand the rationale behind removing the .env files from .gitignore and including a new .env file for Vite configuration. We will follow that approach from now on.

3. **Development Environment Setup:**
   - The instructions for setting up the integrated development environment using Docker Compose are clear. Just to confirm, should we expect any initial setup issues or dependencies that need to be addressed before running the stack?

Overall, these changes seem to improve the development and deployment workflow significantly. I'll proceed with reviewing the code changes in detail.

Thanks.

KianAttar commented:
Hi Amir,

Thank you for your feedback and questions. I’m glad to hear that the changes are looking promising. Here’s a brief overview addressing each of your points:

1. **Dockerfile Addition**:

   The Dockerfile sets up a development environment using Node.js 20 on an Alpine image. It installs dependencies and starts the Vite development server, exposing port 5173. I’ve also added the `src` folder as a volume in Docker Compose to ensure that changes are applied instantly within the container.

   Additionally, I’ve made changes to the Vite configuration file to support development within Docker:

   ```javascript
   server: {
     host: '0.0.0.0',
     port: 5173,
     hmr: {
       host: "localhost",
     }
   }
   ```

   This setup allows the development server to be accessed from outside the container and ensures proper Hot Module Replacement (HMR) functionality. The necessary Nginx proxy rules have been added and tested to support this configuration.

2. **Updated .env File Handling**:

   I’m glad you found the update to the .env file handling clear. As mentioned, we’ve removed the .env files from `.gitignore` and added a new `.env` file for Vite configuration to streamline our environment setup.

3. **Development Environment Setup**:

   To set up the integrated development environment, please create a `.env` file in the root of the project and set the following environment variables:
   - `GOOGLE_CLIENT_ID`
   - `GOOGLE_CLIENT_SECRET`
   - `SESSION_SECRET`

   Docker Compose will use these variables automatically during setup. If you need access to the secrets for the development environment, please send me an email, and I’ll share the details with you.

Feel free to reach out if you have any more questions or need further clarification.
